store/memdb/memdb.go

```diff
@@ -333,3 +351,138 @@ RunLoop:
+		return err
+	}
+
+	// Found existing entry
+	if raw != nil {
+		err = txn.Delete(pipelineRunTable, raw)
+		if err != nil {
+			gaia.Cfg.Logger.Error("failed to delete pipeline run from memdb", "error", err.Error(), "id", runID)
+			return err
+		}
+	}
+
+	// Commit transaction
+	txn.Commit()
+
+	return nil
+}
+
+// InsertDockerWorker inserts a docker worker into the memdb.
+func (m *MemDB) InsertDockerWorker(w *docker.Worker) error {
+	// Create a write transaction
+	txn := m.db.Txn(true)
+
+	// Insert the docker worker object
+	if err := txn.Insert(dockerWorkerTableName, w); err != nil {
+		gaia.Cfg.Logger.Error("failed to insert docker worker via insert", "error", err.Error())
+		return err
+	}
+
+	// Commit transaction
+	txn.Commit()
+
+	return nil
+}
+
+// GetDockerWorker returns the docker worker by the given worker id.
+func (m *MemDB) GetDockerWorker(workerID string) (*docker.Worker, error) {
+	// Create read transaction
+	txn := m.db.Txn(false)
+	defer txn.Abort()
+
+	// Get worker
+	raw, err := txn.First(dockerWorkerTableName, "id", workerID)
+	if err != nil {
+		gaia.Cfg.Logger.Error("failed to get docker worker from memdb", "error", err.Error(), "id", workerID)
+		return nil, err
+	}
+
+	// If nil we couldn't find it
+	if raw == nil {
+		return nil, nil
+	}
+
+	// Convert into docker worker obj
+	w, ok := raw.(*docker.Worker)
+	if !ok {
+		gaia.Cfg.Logger.Error("failed to convert docker worker into docker worker obj", "raw", raw)
+		return nil, errors.New("failed to convert docker worker into docker worker obj")
+	}
+
+	return w, nil
+}
+
+// DeleteDockerWorker deletes the docker worker by then given worker id.
+func (m *MemDB) DeleteDockerWorker(workerID string) error {
+	// Create write transaction
+	txn := m.db.Txn(true)
+
+	// Find existing entry
+	raw, err := txn.First(dockerWorkerTableName, "id", workerID)
+	if err != nil {
+		gaia.Cfg.Logger.Error("failed to load docker worker from memdb", "error", err.Error(), "id", workerID)
+		return err
+	}
+
+	// Found existing entry
+	if raw != nil {
+		err = txn.Delete(dockerWorkerTableName, raw)
+		if err != nil {
+			gaia.Cfg.Logger.Error("failed to delete docker worker from memdb", "error", err.Error(), "id", workerID)
+			return err
+		}
+	}
+
+	// Commit transaction
+	txn.Commit()
+
+	return nil
+}
+
+// GetAllDockerWorker gets all docker worker.
+func (m *MemDB) GetAllDockerWorker() ([]*docker.Worker, error) {
+	var dockerWorkers []*docker.Worker
+
+	// Create a read-only transaction
+	txn := m.db.Txn(false)
+	defer txn.Abort()
+
+	// Get all objects from the docker worker table
+	iter, err := txn.Get(dockerWorkerTableName, "id_prefix")
+	if err != nil {
+		gaia.Cfg.Logger.Error("failed to get docker worker objects from memdb via getalldockerworker", "error", err.Error())
+		return dockerWorkers, nil
+	}
+
+	// Iterate through all items and add them
+	for {
+		item := iter.Next()
+
+		if item == nil {
+			break
+		}
```

Comment:
Why not put this check into the for? 

```go
for item := iter.Next(); item != nil {
```

Reply:
Good catch! I will change that! 😄

Reply:
That actually doesn't work. `item := iter.Next()` is only executed once before loop iteration.

Reply:
true I means to say `item := node; item != nil item = iter.Next`.